Orange/widgets/unsupervised/owtsne.py

```diff
@@ -1,60 +1,235 @@
+            )
+            task.iterations_done += step_size
+            state.set_partial_result(("tsne_embedding", task))
+            if progress_callback is not None:
+                progress_callback(task.iterations_done / total_iterations_needed)
+
+            if state.is_interruption_requested():
+                return
+
+        # Run regular optimization phase
+        while task.iterations_done < total_iterations_needed:
+            # Step size can't be larger than the remaining number of iterations
+            step_size = min(_STEP_SIZE, total_iterations_needed - task.iterations_done)
+            task.tsne_embedding = task.tsne_embedding.optimize(
+                step_size,
+                exaggeration=tsne.exaggeration,
+                momentum=0.8,
+                inplace=False,
+            )
+            task.iterations_done += step_size
+            state.set_partial_result(("tsne_embedding", task))
+            if progress_callback is not None:
+                progress_callback(task.iterations_done / total_iterations_needed)
 
-    @property
-    def n_samples(self):
-        return self.embedding.embedding_.shape[0]
+            if state.is_interruption_requested():
+                return
 
-    def run_optimization(self):
-        total_iterations = self.early_exagg_iter + self.n_iter
+    @classmethod
+    def run(cls, task, state):
+        # type: (Task, TaskState) -> Task
 
-        # Default values of early exaggeration phase
-        exaggeration, momentum = 12, 0.5
+        # Assign weights to each job indicating how much time will be spent on each
+        weights = {"pca": 1, "init": 1, "aff": 23, "tsne": 75}
+        total_weight = sum(weights.values())
 
-        current_iter = self.iterations_done
-        while not current_iter >= total_iterations:
-            # Switch to normal regime if early exaggeration phase is over
-            if current_iter >= self.early_exagg_iter:
-                exaggeration, momentum = self.exaggeration, 0.8
+        # Prepare the tsne object and add it to the spec
+        task.tsne = prepare_tsne_obj(
+            task.data, task.perplexity, task.multiscale, task.exaggeration
+        )
 
-            # Resume optimization for some number of steps
-            self.embedding.optimize(
-                self.step_size, inplace=True, exaggeration=exaggeration,
-                momentum=momentum,
+        # Only execute jobs if needed
+        job_queue = []
+        if task.pca_projection is None:
+            job_queue.append(
+                (partial(cls.compute_pca, task, state), weights["pca"])
+            )
+        if task.initialization is None:
+            job_queue.append(
+                (partial(cls.compute_initialization, task, state), weights["init"])
             )
+        if task.affinities is None:
+            job_queue.append(
+                (partial(cls.compute_affinities, task, state), weights["aff"])
+            )
```

Comment:
How about something like this instead?
```
tasks = [(task.pca_projection, cls.compute_pca, weights["pca"]),
               (task.initialization, cls.compute_initialization, weights["init"]),
               (task.affinities, cls.compute_affinities, weights["aff"])]
job_queue = [(partial(y, task, state), w) for x, y, w in tasks if x]

```

Reply:
This is definitely shorter and much nicer to look at. However, this now seems a little complicated for what is basically several if statements and it takes a couple of seconds to see what's going on. May I propose a very straightforward alternative:

```python
        job_queue = []
        # Add the tasks that still need to be run to the job queue
        if task.pca_projection is None:
            job_queue.append((cls.compute_pca, weights["pca"]))

        if task.initialization is None:
            job_queue.append((cls.compute_initialization, weights["init"]))

        if task.affinities is None:
            job_queue.append((cls.compute_affinities, weights["aff"]))

        total_iterations = task.tsne.early_exaggeration_iter + task.tsne.n_iter
        if task.tsne_embedding is None or task.iterations_done < total_iterations:
            job_queue.append((cls.compute_tsne, weights["tsne"]))

        job_queue = [(partial(f, task, state), w) for f, w in all_jobs]
```

I do like your version, but I think this one will take a couple of seconds less to understand when one of us inevitably returns to this code.